src/source/CHOPCONF.cpp

```diff
@@ -161,3 +161,16 @@ uint8_t TMC2660Stepper::hdec() 	{ GET_REG_2660(hdec);	}
 bool TMC2660Stepper::rndtf()	{ GET_REG_2660(rndtf);	}
 bool TMC2660Stepper::chm()		{ GET_REG_2660(chm);	}
 uint8_t TMC2660Stepper::tbl()	{ GET_REG_2660(tbl);	}
+
+// Explicit instantiation of all possible templates (linker will remove unused ones)
+template class TMCStepper<TMC2130Stepper>;
+template class TMCStepper<TMC5130Stepper>;
+template class TMCStepper<TMC5160Stepper>;
+template class TMCStepper<TMC2208Stepper>;
+template class TMCStepper<TMC2224Stepper>;
+template class TMC2130StepperBase<TMC2130Stepper>;
+template class TMC5130StepperBase<TMC5130Stepper>;
+template class TMC5160StepperBase<TMC5160Stepper>;
+template class TMC2208StepperBase<TMC2208Stepper>;
+template class TMC2208StepperBase<TMC2224Stepper>;
```

Comment:
@ejtagle — It seems like these could be declared in the main header…

Reply:
Unfortunately not... The explicit template instantiation will only instantiate the functions whose definitions (not declarations!) are visible at the time the explicit instantiation is found.
That is the problem with splitting template definitions in several CPP files... :(
The other alternative would be to make the whole library a header...

Reply:
How about if we copy these lines into a separate header and include it in all the places where it's now literally included?

Reply:
That would work @thinkyhead 👍

Reply:
I ended up doing the include at the top without issue. Why would including at the bottom make a difference?

Reply:
Supposedly, the compiler only instantiates the definitions of the functions that are already visible.
Perhaps they added a 2nd pass since the last time i checked it.
If Marlin compiles, then everything should be fine